File: android/engines/src/main/kotlin/com/yizhidao/LineValue.kt
package com.yizhidao

/** Coin / yarrow line values: 6 old yin, 7 young yang, 8 young yin, 9 old yang. */
enum class LineValue(val rawValue: Int) {
    OLD_YIN(6),
    YOUNG_YANG(7),
    YOUNG_YIN(8),
    OLD_YANG(9);

    val isYang: Boolean
        get() = this == YOUNG_YANG || this == OLD_YANG

    val isChanging: Boolean
        get() = this == OLD_YIN || this == OLD_YANG

    val changed: LineValue
        get() = when (this) {
            OLD_YIN -> YOUNG_YANG
            OLD_YANG -> YOUNG_YIN
            YOUNG_YANG, YOUNG_YIN -> this
        }

    /** Binary bit for King Wen lookup: yang = 1, yin = 0. */
    val bit: Int
        get() = if (isYang) 1 else 0

    val displayLabel: String
        get() = when (this) {
            OLD_YANG -> "阳动 9"
            OLD_YIN -> "阴动 6"
            YOUNG_YANG -> "少阳 7"
            YOUNG_YIN -> "少阴 8"
        }

    companion object {
        fun fromRaw(raw: Int): LineValue? = entries.find { it.rawValue == raw }

        fun from(isYang: Boolean, changing: Boolean): LineValue = when {
            isYang && changing -> OLD_YANG
            isYang && !changing -> YOUNG_YANG
            !isYang && changing -> OLD_YIN
            else -> YOUNG_YIN
        }
    }
}

File: ios/Yizhidao/Domain/LineValue.swift
import Foundation

/// Coin / yarrow line values: 6 old yin, 7 young yang, 8 young yin, 9 old yang.
enum LineValue: Int, Codable, Hashable, Sendable {
    case oldYin = 6
    case youngYang = 7
    case youngYin = 8
    case oldYang = 9

    var isYang: Bool {
        switch self {
        case .youngYang, .oldYang: return true
        case .youngYin, .oldYin: return false
        }
    }

    var isChanging: Bool {
        switch self {
        case .oldYin, .oldYang: return true
        case .youngYin, .youngYang: return false
        }
    }

    var changed: LineValue {
        switch self {
        case .oldYin: return .youngYang
        case .oldYang: return .youngYin
        case .youngYang, .youngYin: return self
        }
    }

    /// Binary bit for King Wen lookup: yang = 1, yin = 0.
    var bit: Int { isYang ? 1 : 0 }

    static func from(isYang: Bool, changing: Bool) -> LineValue {
        switch (isYang, changing) {
        case (true, true): return .oldYang
        case (true, false): return .youngYang
        case (false, true): return .oldYin
        case (false, false): return .youngYin
        }
    }
}
